File: app/src/main/java/com/example/kar/horoscope_world/MainActivity.java
package com.example.kar.horoscope_world;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.MenuItem;
import android.widget.Toast;

import com.firebase.ui.auth.AuthUI;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private DrawerLayout drawerLayout;
    private ActionBarDrawerToggle toggle;
    private NavigationView navigationView;

    RecyclerView recyclerView;
    private RecyclerAdapter adapter;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        drawerLayout = findViewById(R.id.drawer);
        toggle = new ActionBarDrawerToggle( this, drawerLayout, R.string.open, R.string.close );
        drawerLayout.addDrawerListener(toggle);
        toggle.syncState();
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
                int id = menuItem.getItemId();
                if ( id == R.id.exit) {
                    AuthUI.getInstance()
                            .signOut(MainActivity.this)
                            .addOnCompleteListener(new OnCompleteListener<Void>() {
                                @Override
                                public void onComplete(@NonNull Task<Void> task) {
                                    Intent intent = new Intent(MainActivity.this, LogIn.class );
                                    startActivity(intent);
                                    finish();
                                }
                            });


                }
                return true;
            }
        });

        FirebaseDatabase firebaseDatabase = FirebaseDatabase.getInstance();
        DatabaseReference databaseReference = firebaseDatabase.getReference("Images");

        ArrayList<Image> imageList = new ArrayList<>();


        int col = 3;
        recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new GridLayoutManager(this, col ));
        adapter = new RecyclerAdapter(this, imageList );
        recyclerView.setAdapter(adapter);

        databaseReference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                ArrayList<Image> imageList = new ArrayList<>();

                for( DataSnapshot snapshot : dataSnapshot.getChildren() ) {
                    Image image = snapshot.getValue(Image.class);
                    imageList.add(image);
                }

                adapter.setImageList(imageList);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                Toast.makeText(MainActivity.this, "Something went wrong", Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected (MenuItem item ) {
        if ( toggle.onOptionsItemSelected(item)) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /*@SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.exit) {
            AuthUI.getInstance()
                    .signOut(this)
                    .addOnCompleteListener(new OnCompleteListener<Void>() {
                        @Override
                        public void onComplete(@NonNull Task<Void> task) {
                            Intent intent = new Intent(MainActivity.this, LogIn.class );
                            startActivity(intent);
                            finish();
                        }
                    });
        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }*/

}

File: app/src/main/java/com/kar/horoscope/world/view/activities/MainActivity.kt
package com.kar.horoscope.world.view.activities

import android.annotation.SuppressLint
import android.os.AsyncTask
import androidx.lifecycle.ViewModelProviders
import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.annotation.RequiresApi
import com.google.android.material.navigation.NavigationView
import androidx.drawerlayout.widget.DrawerLayout
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import android.view.MenuItem
import android.widget.TextView
import com.bumptech.glide.Glide
import com.google.firebase.auth.FirebaseAuth
import com.kar.horoscope.world.R
import com.kar.horoscope.world.repository.MainRepository
import com.kar.horoscope.world.repository.adapter.MainAdapter
import com.kar.horoscope.world.viewmodels.main.MainViewModel
import com.kar.horoscope.world.viewmodels.main.ViewModelFactory
import de.hdodenhof.circleimageview.CircleImageView
import kotlinx.android.synthetic.main.activity_main.*
import org.jsoup.Jsoup
import java.lang.Exception
import java.util.*

class MainActivity : AppCompatActivity() {

    private lateinit var toggle: ActionBarDrawerToggle

    @SuppressLint("CheckResult")
    @RequiresApi(Build.VERSION_CODES.KITKAT)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        val drawer = findViewById<DrawerLayout>(R.id.drawer)
        toggle = ActionBarDrawerToggle(this, drawer, R.string.open, R.string.close )
        drawer.addDrawerListener(toggle)
        toggle.syncState()

        Objects.requireNonNull(supportActionBar)?.setDisplayHomeAsUpEnabled(true)


        ///Set Navigation Header parameters
        val navigationView = findViewById<NavigationView>(R.id.navigationView)
        val view = navigationView.getHeaderView(0)

        val userImage = view.findViewById<CircleImageView>(R.id.userProfileImage)
        val userName = view.findViewById<TextView>(R.id.userName)
        val userEmail = view.findViewById<TextView>(R.id.userEmail)

        val user = FirebaseAuth.getInstance().currentUser
        Glide.with( userImage.context ).load(user?.photoUrl).into(userImage)
        userEmail.text = user?.email
        userName.text = user?.displayName

        val repository = MainRepository( this )
        val viewModel: MainViewModel by lazy {
            ViewModelProviders.of(this,
                ViewModelFactory(repository)
            ).get(MainViewModel::class.java)
        }

        viewModel.getObservableListModel.subscribe{
            val adapter = MainAdapter( it, this )
            recyclerView.layoutManager = GridLayoutManager(this, 3)
            recyclerView.adapter = adapter
        }

        navigationView.setNavigationItemSelectedListener {
            val id = it.itemId
            viewModel.navigationLogic( id )
            true
        }
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if ( toggle.onOptionsItemSelected(item) )
            return true
        return super.onOptionsItemSelected(item)
    }
}
